File: data_manager/preprocess_imgs/create_img_struct_4karpathy.m

%% Load csv with image info
clear all;
target_split = 'train';
fname = sprintf('../../data/fashion53k/csv/imgs/imgs_info_%s.csv', target_split);

% img_id	split	folder	img_filename
fid = fopen(fname, 'rt');
C = textscan(fid, '%s%s%s%s', 'HeaderLines', 1, 'CollectOutput', true, 'delimiter', ',');
fclose(fid);

img_ids = C{1,1}(:,1);
splits = C{1,1}(:,2);
folders = C{1,1}(:,3);
img_filenames = C{1,1}(:,4);

%%
% Load cnn file
n_regions = 4;

fname = sprintf('../../data/fashion53k/img_regions/4_regions_cnn/per_split/cnn_fc7_%s.txt', target_split);

cnn = csvread(fname);

size(cnn);

%% Create structure

Img = cell(length(img_ids), 1);

i = 1; % index over images
j = 1;
k = 1;
bias = ones(n_regions+1,1);
for i = 1:length(img_ids)
    
    s = struct('img_id',[], 'splits', [], 'codes', [], 'folder', [], 'fname', []);
    
    cnn_slice = cnn(j:j+n_regions, :);
    
    s(1).img_id = img_ids{i};
    s(1).splits = splits{i};
    s(1).codes = [cnn_slice bias];
    s(1).folder = folders{i};
    s(1).fname = img_filenames{i};
    
    j = j + n_regions + 1;
    
    Img{k} = s;  
    k = k + 1;
    
end

%%
fname = sprintf('../../data/fashion53k/matlab_structs/split_%s_img_new.mat', target_split);
save(fname, 'Img', '-v7.3')

File: visual_features/sandbox/create_img_struct_4karpathy.py
"""
I tried using this script to create the img struct for karpahty
However, the train file was too for the scipy.io to save.
I ended up writing the script in matlab, which worked a lot better.

Loading the json file in matlab takes way too long.  Perhaps I need another library. So I had to generate the cnn features separately for each split and make a csv file indicating which images belong to which split.

"""
import json
import numpy as np
import scipy.io as sio

print "loading json file"
rpath = '../../data/fashion53k/'
fname = rpath + 'json/dataset_dress_all.json'
with open(fname, 'r') as f:
    data = json.load(f)

print "loading cnn file"
fname = rpath + 'img_regions/4_regions_cnn/cnn_fc7.txt'
f = open(fname, "rb")
cnn = np.loadtxt(f, delimiter=",")
f.close()

print type(cnn)
print cnn.shape

n_regions = 4
cnn_dim = 4096

data_2_matlab_train = []
data_2_matlab_val = []
data_2_matlab_test = []
index0 = 0
bias = np.ones((n_regions + 1, 1))
for item in data['items']:
    img_id = item['imgid']
    split = item['split']
    asin = item['asin']
    folder = item['folder']
    img_filename = item['img_filename']
    # print folder, img_filename
    # if img_id % 1000 == 0:
    #     print img_id

    index1 = index0 + (n_regions + 1)
    print img_id
    # print index0, index1

    cnn_slice = cnn[index0:index1, :]
    print cnn_slice.shape
    # print bias.shape
    cnn_2_store = np.append(cnn_slice, bias, 1)
    # print cnn_slice[:, 0:2]

    d = {}
    d['codes'] = cnn_2_store  # <n_regions x (cnn_dim + 1)>
    d['img_filename'] = img_filename
    d['img_id'] = img_id
    d['split'] = split
    d['asin'] = asin
    d['folder'] = folder

    if split == 'train':
        data_2_matlab_train.append(d)
    elif split == 'val':
        data_2_matlab_val.append(d)
    elif split == 'test':
        data_2_matlab_test.append(d)
    else:
        raise ValueError("must be train, val or test")

    index0 += (n_regions + 1)


print "saving struct file for matlab"
out_fname = rpath + 'matlab_structs/imgs/split_train_img.mat'
sio.savemat(out_fname, {'Img': data_2_matlab_train})

out_fname = rpath + 'matlab_structs/imgs/split_val_img.mat'
sio.savemat(out_fname, {'Img': data_2_matlab_val})

out_fname = rpath + 'matlab_structs/imgs/split_test_img.mat'
sio.savemat(out_fname, {'Img': data_2_matlab_test})
